Further check rangeFromLineNumber parameters

Currently being triggered by the following steps in `linter-eslint`, but stopping it here will fix anywhere else that this is causing crashes on:

1. Try to run `linter-eslint` with a module missing ("parser `babel-eslint` requested but not installed
2. Get a response from the linter of:
```json
{
  "message": "Cannot find module 'babel-eslint'",
  "line": 0,
  "column": 0,
  "severity": 2
}
```
3\. `Helpers.rangeFromLineNumber(textEditor, line - 1)` is then called, meaning `rangeFromLineNumber` is being called with a line of `-1`.

So, tl;dr: Verify that `lineNumber` and `colStart` are within the valid range for a line before attempting to call Atom functions.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,10 +1,11 @@
 ### Upcoming
 
-* Handle column start in `rangeFromLineNumber`, when it is greater than line length 
+* Handle column start in `rangeFromLineNumber`, when it is greater than line length
+* Handle negative column start values and invalid line numbers
 
 ### 3.3.5
 
-* Add Helpers.createElement
+* Add `Helpers.createElement`
 
 ### 3.3.4
 
@@ -16,7 +17,7 @@
 
 ### 3.3.2
 
-* Fix a bug in Helpers.rangeFromLineNumber
+* Fix a bug in `Helpers.rangeFromLineNumber`
 
 ### 3.3.1
 
@@ -32,7 +33,7 @@
 
 ### 3.2.1
 
-* Couple of fixes for findFile
+* Couple of fixes for `findFile`
 * Correct npm `test` script
 
 ### 3.2.0

--- a/lib/helpers.coffee
+++ b/lib/helpers.coffee
@@ -159,7 +159,7 @@ module.exports = Helpers =
     unless names instanceof Array
       names = [names]
     startDir = startDir.split(path.sep)
-    while startDir.length && startDir.join(path.sep)
+    while startDir.length and startDir.join(path.sep)
       currentDir = startDir.join(path.sep)
       for name in names
         filePath = path.join(currentDir, name)

--- a/lib/helpers.coffee
+++ b/lib/helpers.coffee
@@ -159,7 +159,7 @@ module.exports = Helpers =
     unless names instanceof Array
       names = [names]
     startDir = startDir.split(path.sep)
-    while startDir.length
+    while startDir.length and startDir.join(path.sep)
       currentDir = startDir.join(path.sep)
       for name in names
         filePath = path.join(currentDir, name)

--- a/lib/helpers.coffee
+++ b/lib/helpers.coffee
@@ -58,12 +58,11 @@ module.exports = Helpers =
 
   rangeFromLineNumber: (textEditor, lineNumber, colStart) ->
     throw new Error('Provided text editor is invalid') unless textEditor?.getText?
-    if typeof lineNumber isnt 'number' or lineNumber isnt lineNumber or lineNumber < 0 or
-    typeof colStart isnt 'number' or colStart isnt colStart or colStart < 0
+    if typeof lineNumber isnt 'number' or lineNumber isnt lineNumber or lineNumber < 0
       return [[0, 0], [0, 1]]
     lineLength = textEditor.getBuffer().lineLengthForRow(lineNumber)
     throw new Error('Column start greater than line length') if colStart > lineLength
-    unless typeof colStart is 'number'
+    unless typeof colStart is 'number' and not (colStart < 0)
       colStart = (textEditor.indentationForBufferRow(lineNumber) * textEditor.getTabLength())
       if colStart isnt 0
         colStart -= 1

--- a/lib/helpers.coffee
+++ b/lib/helpers.coffee
@@ -44,7 +44,7 @@ module.exports = Helpers =
         spawnedProcess = new BufferedNodeProcess({command, args, options, stdout, stderr, exit})
       else
         spawnedProcess = new BufferedProcess({command, args, options, stdout, stderr, exit})
-      spawnedProcess.onWillThrowError(({error, handle}) =>
+      spawnedProcess.onWillThrowError(({error, handle}) ->
         return reject(error) if error and error.code is 'ENOENT'
         handle()
         if error.code is 'EACCES'
@@ -58,7 +58,8 @@ module.exports = Helpers =
 
   rangeFromLineNumber: (textEditor, lineNumber, colStart) ->
     throw new Error('Provided text editor is invalid') unless textEditor?.getText?
-    if typeof lineNumber is 'undefined' or lineNumber isnt lineNumber
+    if typeof lineNumber isnt 'number' or lineNumber isnt lineNumber or lineNumber < 0 or
+    typeof colStart isnt 'number' or colStart isnt colStart or colStart < 0
       return [[0, 0], [0, 1]]
     lineLength = textEditor.getBuffer().lineLengthForRow(lineNumber)
     throw new Error('Column start greater than line length') if colStart > lineLength

--- a/lib/helpers.coffee
+++ b/lib/helpers.coffee
@@ -62,7 +62,7 @@ module.exports = Helpers =
       return [[0, 0], [0, 1]]
     lineLength = textEditor.getBuffer().lineLengthForRow(lineNumber)
     throw new Error('Column start greater than line length') if colStart > lineLength
-    unless typeof colStart is 'number' and not (colStart < 0)
+    if typeof colStart isnt 'number' or colStart < 0
       colStart = (textEditor.indentationForBufferRow(lineNumber) * textEditor.getTabLength())
       if colStart isnt 0
         colStart -= 1
@@ -159,7 +159,7 @@ module.exports = Helpers =
     unless names instanceof Array
       names = [names]
     startDir = startDir.split(path.sep)
-    while startDir.length and startDir.join(path.sep)
+    while startDir.length
       currentDir = startDir.join(path.sep)
       for name in names
         filePath = path.join(currentDir, name)

--- a/package.json
+++ b/package.json
@@ -23,5 +23,43 @@
   "dependencies": {
     "xregexp": "^2.0.0",
     "tmp": "latest"
+  },
+  "coffeelintConfig": {
+    "max_line_length": {
+      "value": 120,
+      "level": "warn"
+    },
+    "no_empty_param_list": {
+      "level": "error"
+    },
+    "arrow_spacing": {
+      "level": "error"
+    },
+    "no_interpolation_in_single_quotes": {
+      "level": "error"
+    },
+    "no_debugger": {
+      "level": "error"
+    },
+    "prefer_english_operator": {
+      "level": "error"
+    },
+    "colon_assignment_spacing": {
+      "spacing": {
+        "left": 0,
+        "right": 1
+      },
+      "level": "error"
+    },
+    "braces_spacing": {
+      "spaces": 0,
+      "level": "error"
+    },
+    "spacing_after_comma": {
+      "level": "error"
+    },
+    "no_stand_alone_at": {
+      "level": "error"
+    }
   }
 }

--- a/spec/helper-spec.coffee
+++ b/spec/helper-spec.coffee
@@ -176,7 +176,7 @@ describe 'linter helpers', ->
   describe '::createElement', ->
     it 'works', ->
       clicked = false
-      clickListener = () -> clicked = true
+      clickListener = -> clicked = true
 
       el = helpers.createElement('div')
       el.innerHTML = 'Some HTML'

--- a/spec/helper-spec.coffee
+++ b/spec/helper-spec.coffee
@@ -73,6 +73,14 @@ describe 'linter helpers', ->
         atom.workspace.open("#{__dirname}/fixtures/something.js").then ->
           textEditor = atom.workspace.getActiveTextEditor()
           expect(helpers.rangeFromLineNumber(textEditor)).toEqual([[0, 0], [0, 1]])
+          expect(helpers.rangeFromLineNumber(textEditor, -1)).toEqual([[0, 0], [0, 1]])
+          expect(helpers.rangeFromLineNumber(textEditor, 'a')).toEqual([[0, 0], [0, 1]])
+    it 'ignores an invalid starting column', ->
+      waitsForPromise ->
+        atom.workspace.open("#{__dirname}/fixtures/something.js").then ->
+          textEditor = atom.workspace.getActiveTextEditor()
+          expect(helpers.rangeFromLineNumber(textEditor, 1, -1)).toEqual([[1, 0], [1, 40]])
+          expect(helpers.rangeFromLineNumber(textEditor, 1, 'a')).toEqual([[1, 0], [1, 40]])
     it 'returns a range (array) with some valid points', ->
       waitsForPromise ->
         atom.workspace.open("#{__dirname}/fixtures/something.js").then ->

